interfaces/hotplug/spec.go

```diff
@@ -0,0 +1,77 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package hotplug
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/snapcore/snapd/interfaces/utils"
+)
+
+// SlotSpec is a definition of the slot to create in response to udev event.
+type SlotSpec struct {
+	Name  string
+	Label string
+	Attrs map[string]interface{}
+}
+
+type Specification struct {
+	slots map[string]*SlotSpec
+}
+
+func NewSpecification() *Specification {
+	return &Specification{
+		slots: make(map[string]*SlotSpec),
+	}
+}
+
+func (h *Specification) AddSlot(slotSpec *SlotSpec) error {
```

Comment:
This must absolutely associate with a sysfs device. How is this being ensured?

Reply:
This is going to happen outside in the hotplug event processing loop. Hotplug event handler will iterate over all hotplug interfaces, determine device key, create a Specification *per-device*. Interface fills in the Specification by adding slot(s), the outher code knows device key for given set of slots.